nautobot/core/ui/object_detail.py

```diff
@@ -165,417 +161,341 @@ class Tab(ABC):
-                    # TODO: this is ugly - we should abstract this away and provide a standard way of including
-                    # bulk actions and other buttons in the footer
-                    footer_template_string="""\
-{% load perms %}
-{% with request.path|add:"?tab=contacts"|urlencode as return_url %}
-    {% if perms.extras.change_contactassociation %}
-        <button type="submit" name="_edit" formaction="{% url 'extras:contactassociation_bulk_edit' %}?return_url={{return_url}}" class="btn btn-warning btn-xs">
-            <span class="mdi mdi-pencil" aria-hidden="true"></span> Edit
-        </button>
-    {% endif %}
-    {% if perms.extras.delete_contactassociation %}
-        <button type="submit" formaction="{% url 'extras:contactassociation_bulk_delete' %}?return_url={{return_url}}" class="btn btn-danger btn-xs">
-            <span class="mdi mdi-trash-can-outline" aria-hidden="true"></span> Delete
-        </button>
-    {% endif %}
-    {% if perms.extras.add_contactassociation %}
-        <div class="pull-right">
-            <a href="{% url 'extras:object_contact_team_assign' %}?return_url={{return_url}}&associated_object_id={{object.id}}&associated_object_type={{content_type.id}}" class="btn btn-primary btn-xs">
-                <span class="mdi mdi-plus-thick" aria-hidden="true"></span> Add Contact
-            </a>
-        </div>
-    {% endif %}
-    <div class="clearfix"></div>
-{% endwith %}
-""",
-                ),
-            ],
-        )
+        Args:
+            label (str): Label to display for this panel. Optional; if an empty string, the panel will have no label.
+            section (str): One of `nautobot.core.ui.choices.SectionChoices`, indicating which section of the layout
+                this Panel belongs to.
+            body_id (str): HTML element id to attach to the rendered body.
+            body_content_template_path (str): Template path to render the content contained *within* the panel body.
+            header_extra_content_template_path (str): Template path to render extra content into the panel header,
+                if any, not including its label if any.
+            footer_content_template_path (str): Template path to render content into the panel footer, if any.
+            template_path (str): Template path to render the Panel as a whole. Generally you won't override this.
+            body_template_path (str): Template path to render the panel body, including its contents.
+                Generally you won't override this.
+        """
+        self.label = label
+        self.section = section
+        self.body_id = body_id
+        self.body_content_template_path = body_content_template_path
+        self.header_extra_content_template_path = header_extra_content_template_path
+        self.footer_content_template_path = footer_content_template_path
+        self.template_path = template_path
+        self.body_template_path = body_template_path
```

Comment:
Yeah, it's a bit confusing isn't it? Open to suggestions for better names, but basically:

`body_template` is the template for the wrapping `<div class="panel-body">` or `<table class="panel-body">` elements. There should really only be a small number of these templates, all of which are built into the core app.

`body_content_template` is the template for the actual *content* of the panel body, not including that generic outer element. This is the part that will more commonly be customized to various use cases.

I did this split so that the Bootstrap-specific portion (e.g. `panel-body` CSS class and the like) is repeated as few distinct times as possible.

Hope that helps?

Reply:
perhaps instead of `body_template_path` we can do `body_wrapper_template_path`? Not entirely sure tho.

Reply:
Sure, that's probably clearer. :) Will do.